### ISSUE TITLE
Discussion: DispatchAsync for prerendering

### ISSUE THREAD

**Issue by nightBaker** [mrpmorris/Fluxor#146]
I want to use fluxor with prerender ablility. 

In my case, in component onInitilized method dispatches action that has effect for getting model from server. However, Dispatch method is not awaitable, so OnInitialized method thinks that it is done and it returns to client, so I get empty page with loader. 

I suppose, if we have awaitable Dispactch method, OnInitialized would wait and render model on a page.

**Comment by mrpmorris** [comment]
When you await in OnInitializedAsync, Blazor will render before the task completes anyway.

So your render code will need to be able to render before the data is fetched; either by checking if the data is null or by putting an IsLoading member in your state.

**Comment by nightBaker** [comment]
Based on microsoft documentation

Component lifecycle events:

If the component is rendering for the first time on a request:
Create the component's instance.
Perform property injection. Run SetParametersAsync.
**Call OnInitialized{Async}. If a Task is returned, the Task is awaited and the component is rendered. If a Task isn't returned, the component is rendered.**

https://docs.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-5.0

**Comment by mrpmorris** [comment]
The docs are not clear.

It means when control flow is yielded due to an await statement, not when the await completes.

Read the Async section here https://blazor-university.com/components/component-lifecycles/

**Comment by nightBaker** [comment]
You are right, however it relates for rendering on client.
When you use prerender mode, asp.net returns html page after OnInitiliaziedAsync is realy finished.

I tried to await in OnInitializedAsync while It fetch data, but it seems that fluxor is not activated yet.

**Comment by mrpmorris** [comment]
Fluxor initializes after the first render, because it needs to initialize middleware that might need JavaScript. Before that it will queue up the actions.

**Comment by nightBaker** [comment]
So now fluxor does not work with prerender mode.
Do you have any idea how bring it to work ?
Maybe we can init fluxor with middlewares which don't not require JS, firstly ?

**Comment by mrpmorris** [comment]
Fetching of remote data will be Async so will never happen in time for prerender to display the result. And if you know the data at the point of prerender then make it the initial state in the feature.

**Comment by nightBaker** [comment]
_Fetching of remote data will be Async so will never happen in time for prerender to display the result_

This is wrong. 

[This example](https://github.com/nightBaker/examples/tree/master/examples/TestPrerender) shows, that prerender returns page after OnInitiliaziedAsync is finished and data is fetched...

**Comment by mrpmorris** [comment]
I'll investigate that further. 

If you aren't using and middleware that uses JavaScript then you can inject IStore into App.razor and in OnInitializedAsync you can await store.InitializeAsync

Then your actions are dispatched immediately. But Dispatch won't let you await, and I don't think it should. 

Perhaps I could add a task to IStore that gets set when the store has no more effects in execution. I think they should do it, shouldn't it?

**Comment by nightBaker** [comment]
Yes, It should work. However, I think using awaitable dispatcher is more naturally for .net. 

I just don't see reasons why it has to be not awaitable.

**Comment by mrpmorris** [comment]
I'll have to give it some thought after I've finished with lazy loading.

**Comment by mrpmorris** [comment]
Do you intend to await every dispatch?

**Comment by brettwinters** [comment]
I'd have a use for awaitable dispatches too. - for example is  [TypeAhead](https://github.com/Blazored/Typeahead) where the search method is async and the control displays a spinner while waiting for the result..

**Comment by mrpmorris** [comment]
@brettwinters that's not the Flux way.  The key action sets state to show the spinner, an effect tasks to the server,  and the server result is reduced into state to show the type ahead and hide the spinner

**Comment by brettwinters** [comment]
Yes, I know. That's what I've been doing and works for 99% of use cases. 

I don't want to hijack this issue, but it might be useful for everyone to see how to handle this the following senario where we have a search method as follows: 

```c#
private async Task<IEnumerable<Film>> SearchFilms(string searchText) 
    => await httpClient.GetJsonAsync<IEnumerable<Film>>($"https://allfilms.com/api/films/?title={searchText}");
```

Using the flux way we'd dispatch an action then use a while loop with a delay to check until  IsLoaded == true? 

```c#
private async Task<IEnumerable<Film>> SearchFilms(string searchText) 
{
   Dispatcher.Dispatch(new RequestedFilmsAction(...));
   
   //do a while loop / with a delay slow poll until  IsLoaded == true, then continue with result or timeout?
}
```

**Comment by mrpmorris** [comment]
Your effect that calls the server would update state,  and your UI would then automatically update

**Comment by mrpmorris** [comment]
I don't think this can be done in a tidy way. Take the following scenario as an example

1: Dispatch action A
2: EffectA goes to server
3: EffectA dispatches action B with data from server
4: EffectB goes to another server
5: EffectB dispatches action C with data from server

At which point should the await on DispatchAsync complete?

Option 1) Complete after step 3.
Pro: Nice and simple to implement
Con: The chain of execution hasn't actually completed. If your page depends on state from ActionC then the HTML returned from your server will be incomplete

Option 2) Complete after step 5
Pro: The server will definitely return the full set of HTML.
Con: There is no way to know if the dispatch one in a long chain of actions from a single DispatchAsync ,or if they were dispatched as part of one or more other DispatchAsync.

Steps regarding option 2.

1: User clicks Button1
2: Code does `await Dispatcher.DispatchAsync(new ActionA());` - which in total will take 5 seconds.
3: The user decides they don't want to wait and click Button2 which should do something else
4: Code does `await Dispatcher.DispatchAsync(new QuitApplicationAction())` - but this will now have to wait for the earlier task (5 seconds) because the 2nd dispatch will be considered a follow-up task of the first DispatchAsync.

**Comment by mrpmorris** [comment]
I'm going to close this. I don't think there is a logical way to do this.

**Comment by nightBaker** [comment]
Sorry for delayed answer. 

I think option 2 is right way. 

In that way, you don't need to store chain of actions, just use await everytime for dispatching commands and effects. Runtime does it for you. 

Regarding your example of option 2.

1: User clicks Button1
2: Code does await Dispatcher.DispatchAsync(new ActionA()); - which in total will take 5 seconds.
3: The user decides they don't want to wait and click Button2 which should do something else
4: Code does await Dispatcher.DispatchAsync(new QuitApplicationAction()) - it goes in parallel and does not wait for others actions.

**Comment by mrpmorris** [comment]
I doubt think awaiting actions is a good idea,  and different people will want different behaviours.

**Comment by nightBaker** [comment]
Server prerender is important feature for SEO.
Fluxor is awesome, however, this disadvantage does not allow me to use it in production. I think, it can be stop point for many developers. 
Moreover, it takes a lot of time to understand that it will not work with server prerender. I think you should note that in documentation.
